backend/src/mirrors_qa_backend/db/tests.py

```diff
@@ -0,0 +1,148 @@
+from mirrors_qa_backend.settings import Settings
+
+
+@dataclass
+class TestListResult:
+    """Result of query to list tests from the database."""
+
+    nb_tests: int
+    tests: list[models.Test]
+
+
+def filter_test(
+    test: models.Test,
+    *,
+    worker_id: str | None = None,
+    country: str | None = None,
+    status: list[StatusEnum] | None = None,
+) -> bool:
+    """Checks if a test has the same attribute as the provided attribute.
+
+    Base logic for filtering a test from a database.
+    Used by test code to validate return values from list_tests.
+    """
+    if worker_id is not None and test.worker_id != worker_id:
+        return False
+    if country is not None and test.country != country:
+        return False
+    if status is not None and test.status not in status:
+        return False
+    return True
+
+
+def get_test(session: OrmSession, test_id: UUID) -> models.Test | None:
+    return session.scalars(
+        select(models.Test).where(models.Test.id == test_id)
+    ).one_or_none()
+
+
+def list_tests(
+    session: OrmSession,
+    *,
+    worker_id: str | None = None,
+    country: str | None = None,
+    status: list[StatusEnum] | None = None,
+    page_num: int = 1,
+    page_size: int = Settings.MAX_PAGE_SIZE,
+    sort_column: TestSortColumnEnum = TestSortColumnEnum.requested_on,
+    sort_direction: SortDirectionEnum = SortDirectionEnum.asc,
+) -> TestListResult:
+
+    # If no status is provided, populate status with all the allowed values
+    if status is None:
+        status = list(StatusEnum)
+
+    if sort_direction == SortDirectionEnum.asc:
+        direction = asc
+    else:
+        direction = desc
+
+    # By default, we want to sort tests on requested_on. However, if a client
+    # provides a sort_column, we give their sort_column a higher priority
+    order_by: tuple[UnaryExpression[str], ...]
+    if sort_column != TestSortColumnEnum.requested_on:
+        order_by = (
+            direction(sort_column.name),
+            asc(TestSortColumnEnum.requested_on.name),
+        )
+    else:
+        order_by = (direction(sort_column.name),)
+
+    # If a client provides an argument i.e it is not None, we compare the corresponding
+    # model field against the argument, otherwise, we compare the argument to
+    # its default in the database which translates to a SQL true i.e we don't
+    # filter based on this argument.
+    query = (
+        select(func.count().over().label("total_records"), models.Test)
+        .where(
+            (models.Test.worker_id == worker_id) | (worker_id == None),  # noqa
+            (models.Test.country == country) | (country == None),  # noqa
+            (models.Test.status.in_(status)),
+        )
+        .order_by(*order_by)
+        .offset((page_num - 1) * page_size)
+        .limit(page_size)
+    )
+
+    result = TestListResult(nb_tests=0, tests=[])
+
+    for total_records, test in session.execute(query).all():
```

Comment:
What's the use of `total_records` since you're requesting a list of results (`.all()`) and looping over it? Isn't it equivalent to a `len()`?

Reply:
No, actually, `total_records` is the result returned from the SQL window function `func.count().over()`. If I use `len` to capture the length of the records, its value would never be greater than the value that we are limiting the number of rows to.

Assume we apply a filter which has 100 rows in the database matching the criteria, if we used a limit of 25, the value of `len` would never be greater than 25. But because this is a SQL window function, it would return the results over the filters and I can capture it along with the rows returned from the query. It would always return 100 even we set the limit to be 1.

Reply:
Its value is always going to be the same because I included it in the select statement. Thus, storing its value in the actual variable is similar to writing the same value over and over again. But there's no cleaner way I could to think of to do it.

Reply:
OK, my bad then